runtime/compare.c

```diff
@@ -293,16 +297,33 @@ static intnat do_compare_val(struct compare_stack* stk,
   next_item:
     /* Pop one more item to compare, if any */
     if (sp == stk->stack) return EQUAL; /* we're done */
-    v1 = *((sp->v1)++);
-    v2 = *((sp->v2)++);
-    if (--(sp->count) == 0) sp--;
+
+    /* Periodically poll for actions, since this loop can run for
+       unbounded time. */
+    if (++signal_poll_timer >= COMPARE_SIGNAL_POLL_PERIOD) {
+      signal_poll_timer = 0;
+      if (caml_check_pending_actions()) {
+        value exn;
+        Begin_roots_block((value*)(stk->stack),
+                          (stk->limit - stk->stack) * sizeof(struct compare_stack) / sizeof(value));
+        exn = caml_do_pending_actions_exn();
+        End_roots();
+        if (Is_exception_result(exn)) {
+          caml_raise(exn);
```

Comment:
There is a missing `Extract_exception`; or simply call `caml_raise_if_exception`.

Reply:
Or simpler for what you wrote: call `caml_process_pending_actions`. But I suspect you also want to clear the stack before raising.

Reply:
Oops, re-done.